### ISSUE TITLE
Client-side Python fork tests are flaky on Python 3.7

### ISSUE THREAD

**Issue by ericgribkoff** [grpc/grpc#18075]
https://github.com/grpc/grpc/pull/16513 adds fork-without-exec tests for Python clients using `GRPC_ENABLE_FORK_SUPPORT=1` and the `multiprocessing` library. The tests are flaky on Kokoro using Python 3.7, leading to what appear to be hanging child processes.

I have been unable to reproduce the problem locally to attach a debugger and determine the exact problem, but based on a series of trials on Kokoro it looks the child's post-fork Python handler is deadlocking either when [obtaining the GIL](https://github.com/grpc/grpc/blob/04cdb1826666e0de67cfd11c1f5214e0fd53a018/src/python/grpcio/grpc/_cython/_cygrpc/fork_posix.pyx.pxi#L59) and or at some point *before* beginning to [clean-up inherited channels](https://github.com/grpc/grpc/blob/04cdb1826666e0de67cfd11c1f5214e0fd53a018/src/python/grpcio/grpc/_cython/_cygrpc/fork_posix.pyx.pxi#L66). This was determined via inserting calls to C++'s `abort()` or Python's `os._exit` at various points within the post-fork handler: aborting the process before obtaining the GIL deterministically avoided the deadlock, but even simply exiting the process with the GIL (before doing anything potentially suspicious, such as obtaining a condition variable's lock) could still result in a hung child process.

Since it appears that the deadlock is present on Python 3.7 and not on Python 2, and the hang does not appear to occur as a result of an error in our fork-handlers (such as trying to obtain a lock that was held by another thread pre-fork), I suspect that the issue is a result of our pre-fork handlers only *pausing* active gRPC Python threads in [`block_if_fork_in_progress`](https://github.com/grpc/grpc/blob/04cdb1826666e0de67cfd11c1f5214e0fd53a018/src/python/grpcio/grpc/_cython/_cygrpc/fork_posix.pyx.pxi#L113), where threads block on a condition variable until the fork completes. My guess is that these threads, since they are abruptly terminated at fork, are still holding some internal Python lock or otherwise leave the interpreter's state in a broken condition post-fork. I couldn't find a corresponding known issue on the Python issue tracker, nor have I confirmed this to be the case, but it seems the most likely cause at this point - it's also not clear that this would even indicate a bug in Python, as the documentation is unclear as to whether it's even intended to be safe to fork with living (but blocking) threads via the multiprocessing module.

One option to both test the above hypothesis and hopefully fix the underlying issue would be to change gRPC Python's fork handlers to actually stop (`join`) its threads pre-fork, and resume them post-fork. This is the behavior of gRPC core's fork handlers. This is a somewhat invasive change, as unlike core, gRPC Python doesn't have any executor/queue concept for the thread's on-going work to be placed while the thread itself is joined and then re-created in the parent's post-fork handler. A lighter weight approach that may also suffice is to have the threads simply release the GIL prior to fork, block, then reobtain the GIL after fork in the parent.

The [`spawn`](https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods) option added to `multiprocessing` in Python 3.4 may also present an alternative to the current fork-support via fork handlers for Python 3.7 users that would workaround the above issue entirely.

**Comment by ericgribkoff** [comment]
I was able to reproduce the issue in a locally running docker container matching our kokoro test runs. It took repeated runs of the new fork tests for a couple hours to produce the first hang, but artificially increasing CPU load on my desktop seemed to make it much easier to reproduce, which would help explain why it happens so frequently on our CI builds but was hard to reproduce locally.

The backtrace from attaching to one of the hanging processes:

```
#0  futex_wait_cancelable (private=0, expected=0, futex_word=0x8a6bfc <_PyRuntime+1340>)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:88
#1  __pthread_cond_wait_common (abstime=0x0, mutex=0x8a6c00 <_PyRuntime+1344>, cond=0x8a6bd0 <_PyRuntime+1296>)
    at pthread_cond_wait.c:502
#2  __pthread_cond_wait (cond=0x8a6bd0 <_PyRuntime+1296>, mutex=0x8a6c00 <_PyRuntime+1344>)
    at pthread_cond_wait.c:655
#3  0x000000000054ccb1 in drop_gil.lto_priv () at ../Python/ceval_gil.h:183
#4  0x0000000000553fd1 in _PyEval_EvalFrameDefault () at ../Python/ceval.c:979
#5  0x000000000054a8d2 in PyEval_EvalFrameEx (throwflag=0, f=0x7f79f5b785d0) at ../Python/ceval.c:547
#6  _PyEval_EvalCodeWithName () at ../Python/ceval.c:3930
#7  0x00000000005d7750 in _PyFunction_FastCallDict () at ../Objects/call.c:367
#8  0x00000000005910b3 in _PyObject_FastCallDict (kwargs=0x0, nargs=1, args=0x7ffe6be3cbf0, 
    callable=0x7f79f5b38268) at ../Objects/call.c:98
#9  _PyObject_Call_Prepend (kwargs=0x0, args=0x7f79f6308048, obj=<optimized out>, callable=0x7f79f5b38268)
    at ../Objects/call.c:904
#10 slot_tp_init () at ../Objects/typeobject.c:6610
#11 0x0000000000597231 in type_call.lto_priv () at ../Objects/typeobject.c:949
#12 0x00007f79f539a2a4 in __Pyx_PyObject_Call (kw=0x0, arg=0x7f79f6308048, func=0xce4138)
    at src/python/grpcio/grpc/_cython/cygrpc.cpp:68112
#13 __Pyx_PyObject_CallNoArg (func=0xce4138) at src/python/grpcio/grpc/_cython/cygrpc.cpp:3178
#14 0x00007f79f53d2406 in __pyx_f_4grpc_7_cython_6cygrpc___postfork_child ()
    at src/python/grpcio/grpc/_cython/cygrpc.cpp:50541
#15 0x00007f79f6484008 in __run_fork_handlers (who=who@entry=atfork_run_child) at register-atfork.c:134
#16 0x00007f79f6443b9a in __libc_fork () at ../sysdeps/nptl/fork.c:137
#17 0x000000000060794b in os_fork_impl (module=<optimized out>) at ../Modules/posixmodule.c:5438
```

##14 is our Python postfork handler (cygrpc.cpp:50541), and this code is part of the block that corresponds with the following line of [`fork_posix.pyx.pxi`](https://github.com/grpc/grpc/blob/db38d4ca9b7684e9793a8b5edd626ed8d3f76d1b/src/python/grpcio/grpc/_cython/_cygrpc/fork_posix.pyx.pxi#L72):

```
            _fork_state.active_thread_count = _ActiveThreadCount()
```

The stack trace proceeds to [`drop_gil.lto_priv ()` at `../Python/ceval_gil.h:183`](https://github.com/python/cpython/blob/6f352199e4447764bc472e34352d0dff4db8a52d/Python/ceval_gil.h#L183) and the hang occurs.

From the above file's "Notes on the implementation" (of the CPython GIL, which I would think is always a bad sign if your stack trace leads you there...):

>    - When a thread releases the GIL and gil_drop_request is set, that thread
     ensures that another GIL-awaiting thread gets scheduled.
     It does so by waiting on a condition variable (switch_cond) until
     the value of last_holder is changed to something else than its
     own thread state pointer, indicating that another thread was able to
     take the GIL.

That certainly makes it sound like the thread is waiting until another thread grabs the GIL and notifies the condition variable - unfortunately, since we just forked, there is no other thread. It would appear that the hang occurs when the parent process has a Python thread which sets the GIL's `gil_drop_request` to true just before forking, which triggers the hang when the sole surviving thread post-fork attempts to dutifully yield the GIL to the requester.

In conversation with @gnossen today, it came up that our handling of the `_consume_request_iterator` thread in `_channel.py` on fork may not be 100% safe - since this gRPC Python thread may block indefinitely in user-code, we have to ignore it for fork accounting purposes via `cygrpc.enter_user_request_generator()` and `cygrpc.return_from_user_request_generator()`, so this thread may not be actually blocking on anything - other than user-provided code - at fork time. It's possible that this explains the not-quite-always failing tests, if the hang depends on where the consume request iterator has reached when fork is invoked. I will update the tests to guarantee this thread is blocking when fork is invoked and see if that has any impact on the deadlock. However, the test has previously deadlocked on non-bidi-streaming tests, which suggest that the existence of any other threads when fork is called, regardless of whether they are blocking or not, is unsafe.

**Comment by gnossen** [comment]
We're planning on looking into this issue in Q3. We've determined some acceptance criteria:
 - Determine whether cpython 3.8 is affected
 - Understand the nature of the problem and post a summary here. (if any more information is gleaned)
 - (stretch goal) Resolve the issue for all versions of cpython.

**Comment by DrRibosome** [comment]
is there any update on whether this affects cpython 3.8?

**Comment by chenqiuhao1997** [comment]
I have met the same problem on python3.8 with newest grpcio 1.42.0.

The root cause of this problem is that `CPython` reinitialize the `GIL` after `pthread_atfork` child handler and before `os.register_at_fork` child handler in the fork child progress.

In grpcio, the fork child handler is registered into `pthread_atfork`, which means the child handler would run without a correct, reinitialized `GIL`. However, the fork child handler in grpcio uses `GIL` directly (such as `with gil:`) and indirectly (such as `_fork_state.fork_in_progress_condition = threading.Condition()`, which finally causes this problem) by cython.

It is not a good ideal to run python codes in `pthread_atfork` child handler where `GIL` is not ready. Python `os` module provides function `os.register_at_fork` to solve this problem since python3.7. 

An appropriate solution is that grpcio may use `os.register_at_fork` instead of `pthread_atfork` to handle fork on python3.7 or later. There is no simple way to resolve this problem on python3.6 or before unless using magic modifies (such as wrapping `os.fork`) or making a lot of changes to the existing fork handling codes. But even being resolved on python3.7 or later is much better than now.

This is a serious problem because it will never be safe for user to fork since `fork_handlers_and_grpc_init()`is called, even if all grpc requests have been finished and all grpc channels have been closed. Hope this problem could be resolved as soon as possible since this issue has been opened for almost 3 years.

Here are the codes mentioned before.
```python
# grpc/src/python/grpcio/grpc/_cython/_cygrpc/fork_posix.pyx.pxi

cdef void __postfork_child() nogil:
    with gil:
        try:
            if _fork_handler_failed:
                return
            # Thread could be holding the fork_in_progress_condition inside of
            # block_if_fork_in_progress() when fork occurs. Reset the lock here.
            _fork_state.fork_in_progress_condition = threading.Condition()
            # ... ...

def fork_handlers_and_grpc_init():
    grpc_init()
    if _GRPC_ENABLE_FORK_SUPPORT:
        with _fork_state.fork_handler_registered_lock:
            if not _fork_state.fork_handler_registered:
                pthread_atfork(&__prefork, &__postfork_parent, &__postfork_child)
                _fork_state.fork_handler_registered = True

```
```cpp
// cpython/Modules/posixmodule.c

static PyObject *
os_fork_impl(PyObject *module)
{
    pid_t pid;
    // ... ...
    PyOS_BeforeFork();
    pid = fork();
    if (pid == 0) {
        /* child: this clobbers and resets the import lock. */
        PyOS_AfterFork_Child();
    } else {
        /* parent: release the import lock. */
        PyOS_AfterFork_Parent();
    }
    if (pid == -1)
        return posix_error();
    return PyLong_FromPid(pid);
}

void
PyOS_AfterFork_Child(void)
{
    PyStatus status;
    _PyRuntimeState *runtime = &_PyRuntime;
    // ... ...
    status = _PyEval_ReInitThreads(tstate);
    if (_PyStatus_EXCEPTION(status)) {
        goto fatal_error;
    }
    // ... ...
    run_at_forkers(tstate->interp->after_forkers_child, 0);
    return;

fatal_error:
    Py_ExitStatusException(status);
}
```
```cpp
// cpython/Python/ceval.c

PyStatus
_PyEval_ReInitThreads(PyThreadState *tstate)
{
    // ... ...
    if (!gil_created(gil)) {
        return _PyStatus_OK();
    }
    recreate_gil(gil);

    take_gil(tstate);

    // ... ...
}
```

Here are details of my blocked progress.
```
~$ pip3 show grpcio
Name: grpcio
Version: 1.42.0
Summary: HTTP/2-based RPC framework
Home-page: https://grpc.io
Author: The gRPC Authors
Author-email: grpc-io@googlegroups.com
License: Apache License 2.0
Location: /usr/local/lib/python3.8/dist-packages

~$ python3 --version
Python 3.8.10

(gdb) py-bt
Traceback (most recent call first):
  File "/usr/lib/python3.8/threading.py", line 222, in __init__
    def __init__(self, lock=None):
  <built-in method fork of module object at remote 0x7f25d0614360>
  File "/usr/lib/python3.8/multiprocessing/popen_fork.py", line 70, in _launch
    self.pid = os.fork()
  File "/usr/lib/python3.8/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(process_obj)
  File "/usr/lib/python3.8/multiprocessing/context.py", line 277, in _Popen
    return Popen(process_obj)
  File "/usr/lib/python3.8/multiprocessing/context.py", line 224, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "/usr/lib/python3.8/multiprocessing/process.py", line 633, in start
  File "testlocatefork.py", line 51, in <module>

(gdb) bt
#0  futex_wait_cancelable (private=<optimized out>, expected=0,
    futex_word=0x93dba0 <_PyRuntime+1312>)
    at ../sysdeps/nptl/futex-internal.h:183
#1  __pthread_cond_wait_common (abstime=0x0, clockid=0,
    mutex=0x93dba8 <_PyRuntime+1320>, cond=0x93db78 <_PyRuntime+1272>)
    at pthread_cond_wait.c:508
#2  __pthread_cond_wait (cond=0x93db78 <_PyRuntime+1272>,
    mutex=0x93dba8 <_PyRuntime+1320>) at pthread_cond_wait.c:638
#3  0x000000000068d6ef in drop_gil (ceval=0x93d8c8 <_PyRuntime+584>,
    tstate=0x1a82f10) at ../Python/ceval_gil.h:176
#4  0x0000000000573ddf in _PyEval_EvalFrameDefault (f=<optimized out>,
    throwflag=<optimized out>) at ../Python/ceval.c:1242
#5  0x000000000056a0ba in PyEval_EvalFrameEx (throwflag=0,
    f=Frame 0x7f25d053bb20, for file /usr/lib/python3.8/threading.py, line 222, in __init__ (self=<Condition at remote 0x7f25ae333580>, lock=None))
    at ../Python/ceval.c:741
#6  _PyEval_EvalCodeWithName (_co=<optimized out>, globals=<optimized out>,
    locals=<optimized out>, args=<optimized out>, argcount=<optimized out>,
    kwnames=<optimized out>, kwargs=0x7ffceff14088, kwcount=<optimized out>,
    kwstep=1, defs=0x7f25d04afcb8, defcount=1, kwdefs=0x0, closure=0x0,
    name='__init__', qualname='Condition.__init__') at ../Python/ceval.c:4298
#7  0x000000000059d1bc in _PyFunction_Vectorcall (kwnames=0x0, nargsf=1,
    stack=0x7ffceff14080, func=<function at remote 0x7f25d044bdc0>)
    at ../Objects/call.c:436
--Type <RET> for more, q to quit, c to continue without paging--
#8  _PyObject_FastCallDict (kwargs=<unknown at remote 0x7ffceff14080>,
    nargsf=1, args=0x7ffceff14080,
    callable=<function at remote 0x7f25d044bdc0>) at ../Objects/call.c:96
#9  _PyObject_Call_Prepend (kwargs=<unknown at remote 0x7ffceff14080>,
    args=<function at remote 0x7f25d044bdc0>, obj=<unknown at remote 0x1>,
    callable=<function at remote 0x7f25d044bdc0>) at ../Objects/call.c:888
#10 slot_tp_init (self=self@entry=<Condition at remote 0x7f25ae333580>,
    args=args@entry=(), kwds=kwds@entry=0x0) at ../Objects/typeobject.c:6790
#11 0x00000000005a73b7 in type_call (type=<optimized out>, args=(), kwds=0x0)
    at ../Objects/typeobject.c:994
#12 0x00007f25ce19df8e in ?? ()
   from /usr/local/lib/python3.8/dist-packages/grpc/_cython/cygrpc.cpython-38-x86_64-linux-gnu.so
#13 0x00007f25ce202c6f in ?? ()
   from /usr/local/lib/python3.8/dist-packages/grpc/_cython/cygrpc.cpython-38-x86_64-linux-gnu.so
#14 0x00007f25d0c05cd9 in __run_fork_handlers (who=who@entry=atfork_run_child,
    do_locking=do_locking@entry=true) at register-atfork.c:133
#15 0x00007f25d0c5414f in __libc_fork () at ../sysdeps/nptl/fork.c:129
#16 0x000000000063f007 in os_fork_impl (module=<optimized out>)
    at ../Modules/posixmodule.c:6250
#17 os_fork (module=<optimized out>, _unused_ignored=<optimized out>)
    at ../Modules/clinic/posixmodule.c.h:2750
```

**Comment by devinmatte** [comment]
Also experiencing deadlock issues with Python 3.8 and grpcio 1.42.0, had to revert back to 1.30.0 in the meantime

**Comment by tgpfeiffer** [comment]
I'm experiencing this problem when using gRPC in a Jupyter notebook.

I am doing something like
```python
! ls
# do some gRPC calls
! ls
```
and see the following:

- *Without* `GRPC_ENABLE_FORK_SUPPORT`, the second `ls` shows

      E1110 06:52:26.082275717 32361 backup_poller.cc:133] Run client channel backup poller: {"created":"@1636527146.082226642","description":"pollset_work","file":"src/core/lib/iomgr/ev_epollex_linux.cc","file_line":321,"referenced_errors":[{"created":"@1636527146.082224310","description":"Bad file descriptor","errno":9,"file":"src/core/lib/iomgr/ev_epollex_linux.cc","file_line":957,"os_error":"Bad file descriptor","syscall":"epoll_wait"}]} Fe2O3.cif __pycache__/ surfeditor.py

  and I think as a side-effect reconnecting to channels after a long idle period may stop working. (Which is why I am looking into this in the first place.)
- *With* `GRPC_ENABLE_FORK_SUPPORT` I experience the same hanging subprocess, presumably after `fork()` because the process tree looks like

       jovyan   15428  5.9  4.6 1552604 376360 ?      Ssl  08:16   0:04 /usr/local/bin/python -m ipykernel_launcher -f /home/jovyan/.local/share/jupyter/runtime/kernel-57007278-cbad-4812-8c7d-64cad82b50af.json
       jovyan   15952  0.0  4.1 1542360 337556 ?      Sl   08:17   0:00  \_ /usr/local/bin/python -m ipykernel_launcher -f /home/jovyan/.local/share/jupyter/runtime/kernel-57007278-cbad-4812-8c7d-64cad82b50af.json

This is on Python 3.7.13 and grpc 1.46.3.

**Comment by ravwojdyla** [comment]
Experience this with python 3.10 and grpcio 1.49.1. 

Here's a stacktrace of the child process:

```
Thread 3153889 (idle): "MainThread"
    syscall (libc-2.31.so)
    absl::lts_20220623::synchronization_internal::Waiter::Wait (libabsl_synchronization.so.2206.0.0)
    AbslInternalPerThreadSemWait_lts_20220623 (libabsl_synchronization.so.2206.0.0)
    absl::lts_20220623::CondVar::WaitCommon (libabsl_synchronization.so.2206.0.0)
    gpr_cv_wait (libgpr.so.27.0.0)
    stop_threads (libgrpc.so.27.0.0)
    grpc_shutdown_internal_locked (libgrpc.so.27.0.0)
    grpc_shutdown (libgrpc.so.27.0.0)
    cygrpc__close (grpc/_cython/cygrpc.cpython-310-x86_64-linux-gnu.so)
    close_on_fork (grpc/_cython/cygrpc.cpython-310-x86_64-linux-gnu.so)
    _close_on_fork (grpc/_channel.py:1557)
    __Pyx__PyObject_CallOneArg (grpc/_cython/cygrpc.cpython-310-x86_64-linux-gnu.so)
    cygrpc___postfork_child (grpc/_cython/cygrpc.cpython-310-x86_64-linux-gnu.so)
    0x7f0c7fff7293 (libc-2.31.so)
    fork (libc-2.31.so)
    os_fork (python3.10)
    _launch (multiprocessing/popen_fork.py:66)
    __init__ (multiprocessing/popen_fork.py:19)
    _Popen (multiprocessing/context.py:281)
    _Popen (multiprocessing/context.py:224)
    start (multiprocessing/process.py:121)
    <module> (<stdin>:1)
```

Strace shows:

> futex(0x7f0be9ed0940, FUTEX_WAIT_PRIVATE, 0, NULL

I can reproduce this in a deterministic way but it's using pieces of closed source, so would need to work on a minimal reproducible open source code. In the meantime I'm happy to provide some diagnostics on the code if you tell me what you need.

**Comment by XuanWang-Amos** [comment]
@gnossen is this related to the thread hanging issue you've been working on?

**Comment by XuanWang-Amos** [comment]
The issues found in python 3.10 doesn't looks like the same one with the original issue discussed here and likely already fixed during our recent effort to fix fork support.

The original issue is specific to python 3.7, we'll probably close this issue after we drop support for Python 3.7 this year.
